Lib/test/test_capi/test_file.py

```diff
@@ -1,22 +1,67 @@
 import os
 import unittest
-from test.support import import_helper
+from test import support
+from test.support import import_helper, os_helper
 
 _testcapi = import_helper.import_module('_testcapi')
 
 
 class CAPIFileTest(unittest.TestCase):
     def test_py_fopen(self):
+        # Test Py_fopen() and Py_fclose()
+        class FSPath:
+            def __init__(self, path):
+                self.path = path
+            def __fspath__(self):
+                return self.path
+
+        with open(__file__, "rb") as fp:
+            source = fp.read()
+
         for filename in (__file__, os.fsencode(__file__)):
             with self.subTest(filename=filename):
-                content = _testcapi.py_fopen(filename, "rb")
-                with open(filename, "rb") as fp:
-                    self.assertEqual(fp.read(256), content)
+                data = _testcapi.py_fopen(filename, "rb")
+                self.assertEqual(data, source[:256])
+
+                data = _testcapi.py_fopen(FSPath(filename), "rb")
+                self.assertEqual(data, source[:256])
+
+        for filename in (
+            os_helper.TESTFN,
+            os.fsencode(os_helper.TESTFN),
+            os_helper.TESTFN_UNDECODABLE,
+            os_helper.TESTFN_UNENCODABLE,
+        ):
+            with self.subTest(filename=filename):
+                try:
+                    with open(filename, "wb") as fp:
+                        fp.write(source)
 
+                    data = _testcapi.py_fopen(filename, "rb")
+                    self.assertEqual(data, source[:256])
+                finally:
+                    os_helper.unlink(filename)
+
+        # embedded null character/byte in the filename
+        with self.assertRaises(ValueError):
+            _testcapi.py_fopen("a\x00b", "rb")
+        with self.assertRaises(ValueError):
+            _testcapi.py_fopen(b"a\x00b", "rb")
+
+        # non-ASCII mode failing with "Invalid argument"
+        with self.assertRaises(OSError):
+            _testcapi.py_fopen(__file__, "\xe9")
```

Comment:
`"\xe9"` is encoded to `b'\xc3\xa9'`. Please test also non-UTF-8 bytes.

Reply:
It's not possible to pass non-UTF-8 bytes, `PySys_Audit()` decodes the mode from UTF-8 in strict mode:

```c
    if (PySys_Audit("open", "Osi", path, mode, 0) < 0) {
        return NULL;
    }
```

I don't think that it's worth to "support" non-UTF-8 just for the test, whereas it's rejected anyway by `fopen()`.